### PR TITLE
Post and pre order vectorization, vectorized initializers

### PR DESCRIPTION
**Pull request by campsd**

### CODE REVIEW

**Comment by campsd** [comment]
I'm not sure if we would want to expose the `post` order flag to the users in factorize. @yhtang  what do you think?

File: funfact/initializers.py [hunk 1]
```diff
@@ -125,3 +125,22 @@ def __call__(self, shape):
         shape = as_tuple(shape)
         std = (self.scale / shape[self.axis])**0.5
         return std * self.distribution(shape)
+
+
+def vectorize_initializer(initializer, post: bool = True):
```
[the remark this answers is not in the document]
Comment (yhtang) [reply funfact/initializers.py:130]: Also, 'vectorize_initializer' might be a bit too wordy to type. I understand that this is to avoid conflict with the `vectorize` method for the tensor expressions. How about we steal `vmap` from JAX? Or maybe call it `stack` since we are literally stacking up lower dimensionality outputs?

File: funfact/lang/_terminal.py [hunk 2]
```diff
@@ -165,13 +166,14 @@ def __init__(self, *size, symbol=None, initializer=None, optimizable=True):
         self.initializer = initializer
         self.optimizable = optimizable
 
-    def vectorize(self, n):
+    def vectorize(self, n, post):
```
Comment (yhtang) [inline funfact/lang/_terminal.py:169]: As above, maybe change `post` to `axis`?

File: funfact/initializers.py [hunk 3]
```diff
@@ -125,3 +125,22 @@ def __call__(self, shape):
...
+    '''Vectorizes an initializer.
+
+    Args:
+        post (bool):
+            If True, the last index of shape is considered the vectorizing
+            index. If False, the first index of shape tuple is considered
+            the vectorizing index.
+    '''
+    if initializer is None:
```
Comment (yhtang) [inline funfact/initializers.py:139]: I understand that this is to handle the case when an abstract tensor has `None` as the initializer.

However, I think it might be semantically more straightforward if this is handled at the terminal level (AbstractTensor), since vectorizing None is not a very meaningful action.

What we can do is to simply do nothing in `AbstractTensor.vectorize` if the initializer is None, because in that case it'll default to `Normal`, which does not need vectorization.

File: funfact/initializers.py [hunk 4]
```diff
@@ -125,3 +125,22 @@ def __call__(self, shape):
...
+        return None
+
+    def wrapper(shape):
+        return ab.stack([initializer(shape[:-1] if post else shape[1:]) for i
```
Comment (yhtang) [inline funfact/initializers.py:143]: It's better to factor our some calculations before we invoke list comprehension, as if it is too long then we lose the readability benefit.

File: funfact/lang/interpreter/_vectorize.py [hunk 5]
```diff
@@ -47,17 +47,20 @@ class LeafVectorizer(_VectorizerBase):
 
     as_payload = TranscribeInterpreter.as_payload
 
-    def __init__(self, replicas: int, vec_index: P.index):
+    def __init__(self, replicas: int, vec_index: P.index, post: bool = True):
```
Comment (yhtang) [inline funfact/lang/interpreter/_vectorize.py:50]: Maybe `post` -> `axis`?

File: funfact/lang/interpreter/_vectorize.py [hunk 6]
```diff
@@ -84,8 +88,14 @@ def ein(
         self, lhs: P.Numeric, rhs: P.Numeric, precedence: int, reduction: str,
         pairwise: str, outidx: Optional[P.indices], live_indices, **kwargs
     ):
+        if self.post:
```
Comment (yhtang) [inline funfact/lang/interpreter/_vectorize.py:91]: Since the branching is strictly binary here, I think it is better to use `else` for the other return statement.

Comment (yhtang) [reply funfact/lang/interpreter/_vectorize.py:91]: And perhaps factor out the common `*[P.index...]` part and only handles the appending/prepending part in the branches?

File: funfact/vectorization.py [hunk 7]
```diff
@@ -9,7 +9,7 @@
 from funfact.model import Factorization
 
 
-def vectorize(tsrex, n):
+def vectorize(tsrex, n, post: bool = True):
```
Comment (yhtang) [inline funfact/vectorization.py:12]: `post` -> `axis`?

**Review by yhtang — CHANGES_REQUESTED** [review]

**Comment by yhtang** [comment]
> I'm not sure if we would want to expose the `post` order flag to the users in factorize. @yhtang what do you think?
> 
> _Edit:_ After some more thought I think that it probably makes sense to change the default to false and let the user change this if required. Having the vectorizing index first easily allows batched application of numerous NumPy functions.

I agree. Since the convention is to prepend the vectorizing dimension, let's just conform to that.

p.s. Maybe this is a sign that I'm no longer as young and rebellious... :grin:

File: funfact/initializers.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (campsd) [reply funfact/initializers.py:130]: What do you think about renaming it to 'append' but keeping it as boolean? Since we only ever have two options I think a boolean makes most sense, or do you think there will ever be a reason to set an intermediate axis?

File: funfact/lang/_terminal.py (hunk 2 above)
Comment (campsd) [reply funfact/lang/_terminal.py]: See reply above, we can change it, but I feel like a boolean is appropriate?

File: funfact/initializers.py (hunk 3 above)
Comment (campsd) [reply funfact/initializers.py]: Done.

File: funfact/lang/interpreter/_vectorize.py (hunk 6 above)
Comment (campsd) [reply funfact/lang/interpreter/_vectorize.py]: Done.

File: funfact/initializers.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (yhtang) [reply funfact/initializers.py]: Yeah this sounds like a good choice!

**Review by yhtang — APPROVED** [review]
Nice!
